internal/command/command.go

```diff
@@ -0,0 +1,307 @@
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+)
+
+type RunCommand func(
+	*cobra.Command,
+	[]string,
+	*flow.Project,
+	*services.Services,
+) (Result, error)
+
+type Command struct {
+	Cmd   *cobra.Command
+	Flags interface{}
+	Run   RunCommand
+}
+
+var (
+	FilterFlag      = ""
+	FormatFlag      = ""
+	SaveFlag        = ""
+	RunEmulatorFlag = false
+	HostFlag        = ""
+	LogFlag         = "info"
+	NetworkFlag     = ""
+)
+
+// InitFlags init all the global persistent flags
+func InitFlags(cmd cobra.Command) {
+	cmd.PersistentFlags().StringVarP(
+		&FilterFlag,
+		"filter",
+		"x",
+		FilterFlag,
+		"Filter result values by property name",
+	)
+
+	cmd.PersistentFlags().StringVarP(
+		&HostFlag,
+		"host",
+		"",
+		HostFlag,
+		"Flow Access API host address",
+	)
+
+	cmd.PersistentFlags().StringVarP(
+		&FormatFlag,
+		"output",
+		"o",
+		FormatFlag,
+		"Output format, values (json, inline)",
+	)
+
+	cmd.PersistentFlags().StringVarP(
+		&SaveFlag,
+		"save",
+		"s",
+		SaveFlag,
+		"Save result to a filename",
+	)
+
+	cmd.PersistentFlags().StringVarP(
+		&LogFlag,
+		"log",
+		"l",
+		LogFlag,
+		"Log level verbosity, values (none, error, debug)",
+	)
+
+	cmd.PersistentFlags().BoolVarP(
+		&RunEmulatorFlag,
+		"emulator",
+		"e",
+		RunEmulatorFlag,
+		"Run in-memory emulator",
+	)
+
+	cmd.PersistentFlags().StringSliceVarP(
+		&flow.ConfigPath,
+		"conf",
+		"f",
+		flow.ConfigPath,
+		"Path to flow configuration file",
+	)
+
+	cmd.PersistentFlags().StringVarP(
+		&NetworkFlag,
+		"network",
+		"n",
+		NetworkFlag,
+		"Network from configuration file",
+	)
+}
+
+// addCommand add new command to main cmd
+// and initializes all necessary things as well as take care of errors and output
+// here we can do all boilerplate code that is else copied in each command and make sure
+// we have one place to handle all errors and ensure commands have consistent results
+func (c Command) Add(parent *cobra.Command) {
+	c.Cmd.RunE = func(cmd *cobra.Command, args []string) error {
+		// initialize project but ignore error since config can be missing
+		project, _ := flow.LoadProject(flow.ConfigPath)
+
+		clientGateway, err := createGateway(cmd, project)
+		handleError("Gateway Error", err)
+
+		logger := createLogger()
+
+		service := services.NewServices(clientGateway, project, logger)
+
+		// run command
+		result, err := c.Run(cmd, args, project, service)
+		handleError("Command Error", err)
+
+		// format output result
+		formattedResult, err := formatResult(result)
+		handleError("Result", err)
+
+		// output result
+		err = outputResult(formattedResult)
+		handleError("Output Error", err)
+
+		return nil
+	}
+
+	bindFlags(c)
+	parent.AddCommand(c.Cmd)
+}
+
+// createGateway creates a gateway to be used, defaults to grpc but can support others
+func createGateway(cmd *cobra.Command, project *flow.Project) (gateway.Gateway, error) {
+	// create in memory emulator client
+	if RunEmulatorFlag {
+		return gateway.NewEmulatorGateway(), nil
```

Comment:
Is the emulator gateway fully working? I remember you saying that it was still a WIP.

If it's not fully functional, or we don't have a solution to the persistence problem, then I think it's best to just remove `RunEmulatorFlag` for now and not make it an option

Reply:
I can make it work, I just wanted to make a boilerplate for implementation and implement it in another PR so it doesn't get overblown as it doesn't belong in refactoring

Reply:
I will remove flag tho